org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/JavaLanguageServerPlugin.java

```diff
@@ -173,6 +173,9 @@ public void start(BundleContext bundleContext) throws Exception {
 			preferenceManager = new StandardPreferenceManager();
 			projectsManager = new StandardProjectsManager(preferenceManager);
 		}
+		IEclipsePreferences fDefaultPreferenceStore = DefaultScope.INSTANCE.getNode(JavaManipulation.getPreferenceNodeId());
+		String javaCompletionFavoriteMembers = String.join(":", preferenceManager.getPreferences().getJavaCompletionFavoriteMembers());
+		fDefaultPreferenceStore.put(JavaManipulationPlugin.CODEASSIST_FAVORITE_STATIC_MEMBERS, javaCompletionFavoriteMembers);
```

Comment:
Regarding that the completionHandler will pass the favorite prefs to the collector at each completion request, what's benefit to persist it to the preference store?

Reply:
See https://github.com/eclipse/eclipse.jdt.ui/blob/master/org.eclipse.jdt.core.manipulation/common/org/eclipse/jdt/core/manipulation/OrganizeImportsOperation.java#L677

Reply:
I see. Here are new comments:
- Looks like you also need update the store when the preferences changed.
- Since the link you pasted is used to fix the bug https://bugs.eclipse.org/bugs/show_bug.cgi?id=283287, not sure whether the previous workaround in JLS OrganizeImportsHandler is still needed.
https://github.com/eclipse/eclipse.jdt.ls/blob/master/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/OrganizeImportsHandler.java#L101-L106

![image](https://user-images.githubusercontent.com/14052197/90112064-1a014800-dd82-11ea-826b-6c6bbbb53a7a.png)

Reply:
those changes belong to the preference manager

Reply:
When you update the preferenceStore in the plugin start function, the preferences from the client side is not passed to the JLS yet, that means you will always save the default JAVA_COMPLETION_FAVORITE_MEMBERS_DEFAULT to the preferenceStore. If this is not on purpose, then you should do it in Preferences#setJavaCompletionFavoriteMembers

```java
	public Preferences setJavaCompletionFavoriteMembers(List<String> javaCompletionFavoriteMembers) {
		this.javaCompletionFavoriteMembers = (javaCompletionFavoriteMembers == null || javaCompletionFavoriteMembers.isEmpty()) ? JAVA_COMPLETION_FAVORITE_MEMBERS_DEFAULT : javaCompletionFavoriteMembers;
        IEclipsePreferences fDefaultPreferenceStore = DefaultScope.INSTANCE.getNode(JavaManipulation.getPreferenceNodeId());
		fDefaultPreferenceStore.put(JavaManipulationPlugin.CODEASSIST_FAVORITE_STATIC_MEMBERS,String.join(":", this.javaCompletionFavoriteMembers));
		return this;
	}
```

Also i see the upstream code uses ";" to restore the favorite members, but here you use ":" to join the array, is that right?
```java
String[] favourites= pref.split(";");
```

Reply:
Besides, JAVA_COMPLETION_FAVORITE_MEMBERS_DEFAULT in Preferences.java looks a little curious to me. It contains a tailing ":" for each array item. But the default user settings in vscode-java doesn't contain ":". Maybe it works, just curious.

![image](https://user-images.githubusercontent.com/14052197/90122974-b4688800-dd90-11ea-9aaf-1e77cf0af197.png)

Reply:
> Since the link you pasted is used to fix the bug https://bugs.eclipse.org/bugs/show_bug.cgi?id=283287, not sure whether the previous workaround in JLS OrganizeImportsHandler is still needed.

We have to use the JLs workaround because of https://github.com/redhat-developer/vscode-java/issues/1472

Reply:
> When you update the preferenceStore in the plugin start function, the preferences from the client side is not passed to the JLS yet, that means you will always save the default 

We don't use JavaManipulationPlugin.CODEASSIST_FAVORITE_STATIC, but have to set it to the blank string.

Reply:
> Besides, JAVA_COMPLETION_FAVORITE_MEMBERS_DEFAULT in Preferences.java looks a little curious to me. It contains a tailing ":" for each array item. But the default user settings in vscode-java doesn't contain ":". 

Fixed.

Reply:
> Regarding that the completionHandler will pass the favorite prefs to the collector at each completion request, what's benefit to persist it to the preference store?

If the JavaManipulationPlugin.CODEASSIST_FAVORITE_STATIC preference is null, some tests throw NPE at OrganizeImportsOperation.java, line 678.